### ISSUE TITLE
`visitChildrenForSemantics` doesn't respect `BoxyDelegate.children` order

### ISSUE THREAD

**Issue by Albert221** [PixelToast/flutter-boxy#45]
Hey, issue pretty critical for screen reader's users. If I'm `inflate`'ing a widget in the delegate, the inflated widget is in the end of child handles list. For painting and hit-testing I just overriden the `BoxyDelegate.children` to return the `BoxyChild`ren in the order I wanted. But the `visitChildrenForSemantics` doesn't respect this order, it goes using the child handles order, which is not always correct.

Below is a small reproduction example. If you take a look at the app with Accessibility Inspector, you'll see that you cannot really get the button's semantics node, because it's completely obstructed by other semantics node. But only in semantics tree, as the `visitChildrenForSemantics` used the original (wrong) order.

<img width="1277" height="1015" alt="Image" src="https://github.com/user-attachments/assets/f76cf175-35fc-46a0-a590-94b91e6dc71d" />

If I manually edit the `RenderBoxyMixin.visitChildrenForSemantics` to the below, I am able to select the button with semantic hit testing (screenshot below)

```dart
  @override
  void visitChildrenForSemantics(RenderObjectVisitor visitor) {
    for (final child in childHandleMap.values.toList().reversed) {
      if (!child._ignore) {
        visitor(child.render);
      }
    }
  }
```

<img width="551" height="1098" alt="Image" src="https://github.com/user-attachments/assets/3dcda53b-7423-43c2-8030-3f23a2d09b64" />

<details>
<summary>Code for reproduction</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:boxy/boxy.dart';

void main() {
  runApp(MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: BodyAndFooter(
          bodyBuilder:
              (context, footerHeight) => ColoredBox(
                color: Colors.greenAccent,
                child: ListView(
                  children: [
                    Text('text1'),
                    Placeholder(),
                    Text('text2'),
                    Placeholder(),
                    Text('text3'),
                    SizedBox(height: footerHeight),
                  ],
                ),
              ),
          footer: Container(
            padding: const EdgeInsets.all(16),
            color: Colors.red.withValues(alpha: 0.5),
            child: ElevatedButton(
              onPressed: () {
                debugDumpSemanticsTree(DebugSemanticsDumpOrder.inverseHitTest);
              },
              child: Text('Footer'),
            ),
          ),
        ),
      ),
    );
  }
}


class BodyAndFooter extends StatelessWidget {
  const BodyAndFooter({
    super.key,
    required this.bodyBuilder,
    required this.footer,
  });

  final Widget Function(BuildContext context, double footerHeight) bodyBuilder;
  final Widget footer;

  @override
  Widget build(BuildContext context) {
    return CustomBoxy(
      delegate: _BodyAndFooterDelegate(bodyBuilder: bodyBuilder),
      children: [footer],
    );
  }
}

class _BodyAndFooterDelegate extends BoxyDelegate {
  _BodyAndFooterDelegate({required this.bodyBuilder});

  final Widget Function(BuildContext context, double footerHeight) bodyBuilder;

  BoxyChild get footer => children.first;

  @override
  Size layout() {
    final rect = Offset.zero & constraints.biggest;

    footer.layout(constraints.tighten(width: rect.width));
    footer.positionRect(rect, Alignment.bottomCenter);

    final body = inflate(
      bodyBuilder(buildContext, footer.size.height),
      id: #body,
    );
    body.layoutRect(rect);

    return rect.size;
  }

  @override
  List<BoxyChild> get children => super.children.reversed.toList();

  @override
  bool shouldRelayout(_BodyAndFooterDelegate oldDelegate) =>
      bodyBuilder != oldDelegate.bodyBuilder;
}


```

</details>

For debugging and reproduction, remember that there is some issue with iOS simulator where hot-restart doesn't really update the semantics tree properly. kill the app and run it again after making the change to semantics tree

**Comment by Albert221** [comment]
As for the fixes, can we just use the `delegate.children` for `visitChildrenForSemantics` in the `RenderBoxy`, and also, for the same reason, for `visitChildren`?

**Comment by Albert221** [comment]
@pingbird is the fix metnioned above okay? Do you want to fix it or should I submit a PR?

**Comment by pingbird** [comment]
Re-constructing a list every time `.children` is accessed isn't super performant, I am still looking for a clean way to do this.

**Comment by Kendru98** [comment]
@pingbird Any update on this? 🙏

**Comment by Albert221** [comment]
@pingbird maybe instead of solving the problem from the issue (which is still pending as you're looking for a clean way) we could allow your package's users to solve the problem on their side. The idea I have is to add a `visitChildrenForSemantics` to the `BaseBoxyDelegate`, which by default would visit `children` if `!child.isIgnored`, the same way as [here](https://github.com/pingbird/boxy/blob/617f9555632146a9fa8772bf24549258af5f575c/boxy/lib/src/boxy/custom_boxy_base.dart#L238).

...and because it's a part of the public (protected) `BoxyDelegate` API, users would be able to override it to their liking.

It's very important for us to have this capability, because Boxy allows us to paint the children in any order we like (because we can alter the `children` and/or `paint`/`paintChildren`, but right now, we cannot update the semantics order according to the paint order which is a huge blocker, in case of more complex layouts that need to be accessible.

Please respond 🙏 and happy new year 🎆
